docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,255 @@
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
+    List() ([]Image, error)
+    Pull(image ImageSpec, auth AuthConfig) error
+    Remove(image ImageSpec) error
+    Status(image ImageSpec) (Image, error)
+    Metrics(image ImageSpec) (ImageMetrics, error)
+}
+
+type ContainerMetricsGetter interface {
+    ContainerMetrics(id string) (ContainerMetrics, error)
+}
+
+All functions listed above are expected to be thread-safe.
+```
+
+### Pod Lifecycle
+
+The PodSandbox’s lifecycle is decoupled from the containers, i.e., a sandbox
+is created before any containers, and can exist after all containers in it have
+terminated.
+
+Assume there is a pod with a single container C. To start a pod:
+
+```
+  create sandbox Foo --> create container C --> start container C
+```
+
+To delete a pod:
+
+```
+  stop container C --> remove container C --> delete sandbox Foo
+```
+
+Kubelet is responsible for creating, starting the containers based on the
+pod-level restart policy. E.g., if container C dies unexpectedly, Kubelet
+will simply create and start the container again in sandbox Foo.
+
+Kubelet is also responsible for gracefully terminating all the containers
+in the sandbox before deleting the sandbox. If Kubelet chooses to delete
+the sandbox with running containers in it, those containers should be forcibly
+deleted.
+
+Note that every PodSandbox/container lifecycle operation (create, start,
+stop, delete) should either return an error or block until the operation
+succeeds. A successful operation should include a state transition of the
+PodSandbox/container. E.g., if a `Create` call for a container does not
+return an error, the container state should be "created" when the runtime is
+queried.
+
+### Updates to PodSandbox or Containers
+
+Kubernetes support updates only to a very limited set of fields in the Pod
+Spec.  These updates may require containers to be re-created by Kubelet. This
+can be achieved through the proposed, imperative container-level interface.
+On the other hand, PodSandbox update currently is not required.
+
+
+### Container Lifecycle Hooks
+
+Kubernetes supports post-start and pre-stop lifecycle hooks, with ongoing
+discussion for supporting pre-start and post-stop hooks in
+[#140](https://issues.k8s.io/140).
+
+These lifecycle hooks will be implemented by Kubelet via `Exec` calls to the
+container runtime. This frees the runtimes from having to support hooks
+natively.
+
+Illustration of the container lifecycle and hooks:
+
+```
+            pre-start post-start    pre-stop post-stop
+               |        |              |       |
+              exec     exec           exec    exec
+               |        |              |       |
+ create --------> start ----------------> stop --------> remove
+```
+
+In order for the lifecycle hooks to function as expected, the `Exec` call
+will need access to the container's filesystem (e.g., mount namespaces).
```

Comment:
Providing an example use case for this would be helpful. Generally hooks have been discussed in the context of executing external plugins to say setup or tear down storage, networking, monitoring, etc. But, these hooks happen in the context of a container filesystem so, it seems like a different use case.